File: showly-react-native/src/features/shows/data/repositories/showRepositoryImpl.ts
import { Show } from "../../domain/entities/show";
import { ShowRepository } from "../../domain/repositories/showRepository";
import { TvMazeRemoteDataSource } from "../datasources/tvMazeRemoteDataSource";
import { mapTvMazeShowToEntity } from "../models/showModel";

export class ShowRepositoryImpl implements ShowRepository {
  constructor(private readonly remoteDataSource: TvMazeRemoteDataSource) {}

  async getShows(): Promise<Show[]> {
    const shows = await this.remoteDataSource.getShows();
    return shows.slice(0, 40).map(mapTvMazeShowToEntity);
  }
}

File: showly_flutter/lib/features/shows/data/repositories/show_repository_impl.dart
import '../../domain/entities/show.dart';
import '../../domain/repositories/show_repository.dart';
import '../datasources/tvmaze_remote_datasource.dart';
import '../models/show_model.dart';

class ShowRepositoryImpl implements ShowRepository {
  final TvMazeRemoteDataSource remoteDataSource;

  ShowRepositoryImpl(this.remoteDataSource);

  @override
  Future<List<Show>> getShows() async {
    final response = await remoteDataSource.getShows();

    return response
        .take(40)
        .map((json) => ShowModel(json as Map<String, dynamic>).toEntity())
        .toList();
  }
}
